packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java

```diff
@@ -0,0 +1,222 @@
+    if (clusterManagersToAdd != null) {
+      for (Object clusterToAdd : clusterManagersToAdd) {
+        addClusterManager(clusterToAdd);
+      }
+    }
+  }
+
+  /** Adds new ClusterManager. */
+  void addClusterManager(Object clusterManagerData) {
+    String clusterManagerId = getClusterManagerId(clusterManagerData);
+    if (clusterManagerId == null) {
+      throw new IllegalArgumentException("clusterManagerId was null");
+    }
+    ClusterManager<MarkerBuilder> clusterManager =
+        new ClusterManager<>(context, googleMap, markerManager);
+    ClusterRenderer clusterRenderer = new ClusterRenderer(context, googleMap, clusterManager, this);
+    clusterManager.setRenderer(clusterRenderer);
+    initListenersForClusterManager(clusterManager, this, clusterItemListener);
+    clusterManagerIdToManager.put(clusterManagerId, clusterManager);
+  }
+
+  /** Removes ClusterManagers by IDs. */
+  public void removeClusterManagers(List<Object> clusterManagerIdsToRemove) {
+    if (clusterManagerIdsToRemove == null) {
+      return;
+    }
+    for (Object rawClusterManagerId : clusterManagerIdsToRemove) {
+      if (rawClusterManagerId == null) {
+        continue;
+      }
+      String clusterManagerId = (String) rawClusterManagerId;
+      removeClusterManager(clusterManagerId);
+    }
+  }
+
+  private void removeClusterManager(Object clusterManagerId) {
+    final ClusterManager<MarkerBuilder> clusterManager =
+        clusterManagerIdToManager.remove(clusterManagerId);
+    if (clusterManager == null) {
+      return;
+    }
+    initListenersForClusterManager(clusterManager, null, null);
+    clusterManager.clearItems();
+    clusterManager.cluster();
+  }
+
+  /** Adds item to the ClusterManager it belongs to. */
+  public void addItem(MarkerBuilder item) {
+    ClusterManager<MarkerBuilder> clusterManager =
+        clusterManagerIdToManager.get(item.clusterManagerId());
+    if (clusterManager != null) {
+      clusterManager.addItem(item);
+      clusterManager.cluster();
+    }
+  }
+
+  /** Removes item from the ClusterManager it belongs to. */
+  public void removeItem(MarkerBuilder item) {
+    ClusterManager<MarkerBuilder> clusterManager =
+        clusterManagerIdToManager.get(item.clusterManagerId());
+    if (clusterManager != null) {
+      clusterManager.removeItem(item);
+      clusterManager.cluster();
+    }
+  }
+
+  /** Called when ClusterRenderer has rendered new visible marker to the map. */
+  void onClusterItemMarker(@NonNull MarkerBuilder item, @NonNull Marker marker) {
+    if (clusterItemListener != null) {
+      clusterItemListener.onClusterItemMarker(item, marker);
+    }
+  }
+
+  /** Reads clusterManagerId from object data. */
+  @SuppressWarnings("unchecked")
+  private static String getClusterManagerId(Object clusterManagerData) {
+    Map<String, Object> clusterMap = (Map<String, Object>) clusterManagerData;
+    return (String) clusterMap.get("clusterManagerId");
```

Comment:
Current implementation of the google maps plugin uses custom serialisation for method channel communication.
These strings are hardcoded through the native implementation matching the data structure serialised at the dart layer.

Corresponding mapping for this is done here:
https://github.com/flutter/packages/blob/dce6f0c91fcf01ee3a67e3475155ec9f16ce2a66/packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/cluster_manager.dart#L62

Should I add this link to before the line, as this might get outdated (even with the ref) at some point.
The current communication layer relies a lot of these hardcoded mapping keys.
Moving the plugin method channel over pigeon in the future would improve the type-safety a lot.

Reply:
(Tarrin just recently landed support for channel suffixes in Pigeon, which unblocks converting maps, so I'm hoping to do the conversion this quarter.)